fendermint/vm/topdown/src/vote/mod.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+mod operation;
+
+use crate::sync::TopDownSyncEvent;
+use crate::vote::operation::{OperationMetrics, OperationStateMachine};
+use crate::BlockHeight;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tokio::sync::{broadcast, mpsc};
+
+#[derive(Clone)]
+pub struct VoteRecord {}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Config {
+    /// The reactor request channel buffer size
+    req_channel_buffer_size: usize,
+    /// The number of requests the reactor should process per run before handling other tasks
+    req_batch_processing_size: usize,
+    /// The number of vote recording requests the reactor should process per run before handling other tasks
+    gossip_req_processing_size: usize,
+    /// The time to sleep for voting loop if nothing happens
+    voting_sleep_interval_sec: u64,
+}
+
+pub struct VoteReactorClient {
+    tx: mpsc::Sender<VoteReactorRequest>,
+}
+
+pub fn start_vote_reactor(
+    config: Config,
+    gossip_rx: broadcast::Receiver<VoteRecord>,
+    gossip_tx: mpsc::Sender<VoteRecord>,
+    internal_event_listener: broadcast::Receiver<TopDownSyncEvent>,
+) -> VoteReactorClient {
+    let (tx, rx) = mpsc::channel(config.req_channel_buffer_size);
+
+    tokio::spawn(async move {
+        let sleep = Duration::new(config.voting_sleep_interval_sec, 0);
+
+        let inner = VotingHandler {
+            req_rx: rx,
+            gossip_rx,
+            gossip_tx,
+            internal_event_listener,
+            config,
+        };
+        let mut machine = OperationStateMachine::new(inner);
+        loop {
+            machine = machine.step();
+            tokio::time::sleep(sleep).await;
+        }
+    });
+
+    VoteReactorClient { tx }
```

Comment:
do you have an example that elaborates this more?

Reply:
Something like having some `new()` function, and a separate `run()` function which blocks indefinitley, and should only be run within a background-task runner like `tokio::spawn`. 

They can then be packaged into one function, like currently. but it will help to better define the in/out flow.

Reply:
I think for this simple function, it should be fine? If the initialisation becomes complex, I will convert it to a `new`/`run` methods.